urlc8dbc7ce6e_dlubarov_guava_tgz-pJ8-MainJ8/src/c/exp/InstanceMethodInvocation.java

```diff
@@ -0,0 +1,154 @@
+package c.exp;
+
+import static util.StringUtils.implode;
+
+import java.util.*;
+
+import util.ArrayUtils;
+
+import common.*;
+
+import c.*;
+import c.ty.*;
+import d.Opcodes;
+
+public class InstanceMethodInvocation extends Expression {
+    public final Expression target;
+    public final String methodName;
+    public final Type[] genericArgs;
+    public final Expression[] args;
+
+    public InstanceMethodInvocation(Expression target, String methodName,
+            Type[] genericArgs, Expression[] args) {
+        this.target = target;
+        this.methodName = methodName;
+        this.genericArgs = genericArgs;
+        this.args = args;
+    }
+
+    private MethodDef getMethod(CodeContext ctx) {
+        Type targetType = target.inferType(ctx);
+        Type[] argTypes = Expression.inferAllTypes(args, ctx);
+        Set<MethodDef> options = new HashSet<MethodDef>();
+        for (ParameterizedType concreteSuper : targetType.getConcreteSupertypes(ctx.type, ctx.method))
+            try {
+                TypeDef superDef = Project.singleton.resolve(concreteSuper.rawType);
+                options.add(superDef.getInstanceMethod(methodName, concreteSuper.genericArgs, genericArgs, argTypes, ctx));
+            } catch (NoSuchElementException e) {}
+        if (options.isEmpty())
+            throw new NoSuchElementException(String.format(
+                    "%s has no method named '%s' with argument types %s.",
+                    target, methodName, Arrays.toString(argTypes)));
+        if (options.size() > 1)
+            throw new NiftyException(
+                    "%s inherits multiple methods named '%s'.",
+                    target, methodName);
+        return options.iterator().next();
+    }
+
+    // Tests whether the method being invoked actually exists.
+    public boolean isValid(CodeContext ctx) {
+        try {
+            getMethod(ctx);
+            return true;
+        } catch (NoSuchElementException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public Type inferType(CodeContext ctx) {
+        MethodDef method = getMethod(ctx);
+        ParameterizedType targetAsMethodOwner = target.inferType(ctx).asSupertype(method.owner, ctx);
+        return method.returnType.withGenericArgs(targetAsMethodOwner.genericArgs, genericArgs);
+    }
+
+    private static Integer getNativeOp(MethodDef m) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 40. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 67-72 into a separate method.